README
# Infomation
	Type       : PHP program
	Name       : Convert MySQL to SQLite
	Version    : 2.1
	Update     : 2012-07-26
	Author     : Yuusaku Miyazaki (宮崎 雄策) (toumin.m7@gmail.com)
	Author-URI : http://d.hatena.ne.jp/sutara_lumpur/20120714/1342269933
	Demo       : http://www.usamimi.info/~sutara/sample/convert_mysql_to_sqlite/
	License    : MIT License (http://www.opensource.org/licenses/mit-license.php)


CHANGELOG.md
# convert-mysql-to-sqlite CHANGELOG

## 2.1.1 (2014-08-09)
- Plugin name was changed.  
  (プラグイン名、ファイル名、フォルダ名を変更しました。)

## v2.1 (2012-07-26)
- The error when using "phpMyAdmin" is corrected.  
  (phpMyAdminでダンプしたSQLの変換で、数行とばして処理してしまうことがある不具合を修正しました。)

## v2.0 (2012-07-20)
- You can use the escaped "back-quote" and the escaped "new-line" in database.  
  (エスケープされたバッククオート、改行文字に対応しました。)

## v1.1 (2012-07-17)
- Stability of operation was improved.  
  (動作の安定性を高めました。)

## v1.0 (2012-07-14)
- The source code was exhibited newly.

src/generate.php
<?php
/**
 * 非同期でファイルを少しずつ変換する。
 */
$date = date('YmdHis');
$upfile = $date.'_mysql.sql';
$dlfile = $date.'_sqlite3.sql';
?>
<!doctype html>
<html lang="ja">
	<head>
		<meta charset="utf-8">
		<title></title>
		<script type="text/javascript" src="//code.jquery.com/jquery.min.js"></script>
		<script type="text/javascript">
			var filesize;
			var upfile = '<?php echo $upfile ?>';
			var dlfile = '<?php echo $dlfile ?>';
			function getFileBitByBit(pos) {
				$.getJSON(
					'transfer.php',
					{
						pos: pos,
						upfile: upfile,
						dlfile: dlfile
					},
					function(json) {
						var per = Math.floor(json.pos / filesize * 100);
						$('#msg').val(per + '% complete.\n');

						// 最後まで読み込んでいなければ、再び実行
						if (json.feof) {
							$('#msg').val('100% complete.\nA file is send back.');
							location.href = 'download.php?upfile=' + upfile + '&dlfile=' + dlfile;
						} else {
							getFileBitByBit(json.pos);
						}
					}
				);
			}
			jQuery(document).ready(function($) {
				getFileBitByBit(0);
			});
		</script>
		<style type="text/css">
		</style>
	</head>
	<body>
		<p><a href="../">Back</a></p>
		<?php
			if (is_uploaded_file($_FILES['upfile']['tmp_name'])) {
				if (move_uploaded_file(
					$_FILES['upfile']['tmp_name'],
					$upfile
				)) {
					touch($dlfile);
					chmod($upfile, 0666);
					chmod($dlfile, 0666);
					echo '<script>filesize = '.filesize($upfile).'</script>';
				} else {
					echo 'Failed to send a file.';
				}
			} else {
				echo 'File is not selected.';
			}
		?>
		<textarea id="msg" rows="4" cols="40"></textarea>
	</body>
</html>

src/transfer.php
<?php
/**
 * 実際の変換処理
 */

/**
 * 置換のコールバック関数
 */
function callback_create($matches) {
	if (isset($matches[4])) {
		return '';
	} elseif (isset($matches[3])) {
		return '';
	} elseif (isset($matches[2]) && isset($matches[1])) {
		return $matches[1].'integer';
	}
}

function callback_backquote($matches) {
	if (isset($matches[2])) {
		// SQLite用に、バッククオートをダブルクオートに変換する。
		return '"';
	} elseif (isset($matches[1])) {
		// クオートで囲まれた中にあるバッククオートは変換しない。
		return $matches[1];
	}
}

/**
 * ファイルへの書き込み
 */
function my_fwrite($fp, $line) {
	// バッククオートをダブルクオートに
	$line = preg_replace_callback('/(\'(?:(?:(?!\\\\).)?(?:(?:\\\\\\\\)*\\\\)\'|[^\'])*\')|(`)/us', 'callback_backquote', $line);

	// '\n'を改行に変換する
	$line = preg_replace('/((?:(?!\\\\).)?(?:\\\\\\\\)*)(\\\\n)/us', '\1'."\n", $line);

	fwrite($fp, $line);
}

/**
 * 変換処理
 */
// ファイルを開く
$fp1 = fopen($_GET['upfile'], 'r');
$fp2 = fopen($_GET['dlfile'], 'a');
//JavaScriptへ返す配列を準備
$return = array(
	'feof' => false,
	'pos'  => $_GET['pos']
);
fseek($fp1, $return['pos']);

// 1度に処理するのは100行まで。
// 
for ($i=0; ($line = fgets($fp1)) && ($i < 100); $i++) {
	// INSERT文の処理
	// phpMyAdminによるダンプの場合
	if (preg_match('/^(INSERT INTO \`[^\`]+\` )[\s\S]*VALUES$/ui', $line, $matches)) {
		$i++;
		$line = fgets($fp1);
		
		while (preg_match('/^([\s\S]+),$/u', $line, $matches2)) {
			$line = $matches[1].'VALUES'.$matches2[1].";\n";
			my_fwrite($fp2, $line);
			
			//次のループのための準備
			//※ ループの条件に適さないと、1行余分に読み込んだことになる。
			//このため、CREATE文よりも先にこの処理を記述しなければならない。
			$i++;
			$line = fgets($fp1);
		}
		if (preg_match('/;$/u', $line)) {
			$line = $matches[1].'VALUES'.$line."\n";
			my_fwrite($fp2, $line);
		}
		//INSERT文が終わったら、INSERT文以外の変換作業へ移る。
	}
	// 端末その他によるダンプの場合
	elseif (preg_match('/^(INSERT INTO \`[^\`]+\` VALUES\s?)([\s\S]*\);)$/ui', $line, $matches)) {
		while (preg_match_all('/^(\((?:(?:\'(?:(?:(?!\\\\).)?(?:(?:\\\\\\\\)*\\\\)\'|[^\'])*\'|[0-9]+|NULL),? ?)+\)), ?([\s\S]*)/ui', $matches[2], $matches2)) {
			$matches[2] = $matches2[2];
			$line = $matches[1].$matches2[1].";\n";
			my_fwrite($fp2, $line);
		}
		$line = $matches[1].$matches[2]."\n";
		my_fwrite($fp2, $line);
	}

	// CREATE文の処理
	if (preg_match('/^CREATE TABLE[\s\S]*(\`[^\`]+\`)/ui', $line, $matches)) {
		my_fwrite($fp2, "CREATE TABLE {$matches[1]} (\n");
		
		$i++;
		$line = fgets($fp1);
		
		$j = 0;
		while (!preg_match('/^\) ENGINE=/i', $line)) {

			// int系をintegerへ、AUTO_INCREMENTを削除する
			if (!preg_match('/^\s*UNIQUE KEY/ui', $line)) {
				// 以下の置換で行末のカンマを削除している。CREATE文が続くなら、カンマを復活させる。
				if ($j > 0) fwrite($fp2, ",\n");
				$j++;

				$line = preg_replace_callback('/^(\s*`[^`]+`\s*)(int|tinyint|samllint|mediumint|bigint)(?:\([^\)]+\))?|(AUTO_INCREMENT),?\\n$|(,\\n)$/ui', 'callback_create', $line);
				my_fwrite($fp2, $line);
			}
			$i++;
			$line = fgets($fp1);
		}
		// ") ENGINE=" の文は、ENGINE以降を削除
		fwrite($fp2, "\n);\n");
		$flag_create_table = false;
	}
	$return['pos']  = ftell($fp1);
}
$return['feof'] = feof($fp1);
fclose($fp1);
fclose($fp2);
echo json_encode($return);

index.php
<!doctype html>
<html lang="ja">
	<head>
		<meta charset="UTF-8">
		<title>Convert MySQL to SQLite</title>
		<style type="text/css">
			form {
				border : 1px solid #aaa;
				border-radius : 4px;
			}
			address {
				margin-top:2em;
				border-top:2px solid #486;
				padding-top:8px;
			}
		</style>
	</head>
	<body>
		<h1>Convert MySQL to SQLite</h1>
		<p>Last update : 2012-07-20</p>
		<p>容量制限 : 10MB以内</p>
		<p>ローカルで実行する際の要件 : PHP5, JavaScript</p>
		<p>phpMyAdminや端末でダンプしたMySQLのSQLファイルを、SQLiteでインポートできるように加工します。</p>
		<p>
			重要なデータベースはWeb上で変換せず、ソースコードをダウンロードして、ローカル環境でPHPを動かして使ってください。<br>
			<a href="https://github.com/SutaraLumpur/convert_mysql_to_sqlite/zipball/master">GitHubからダウンロード</a>
		</p>
		<form action="generate.php" method="POST" enctype="multipart/form-data">
			<input type="file" name="upfile">
			<input type="submit" name="submit" value="Submit">
		</form>

		<!--**************************************************** -->

		<address>
			Author : sutara_lumpur /
			<a href="http://d.hatena.ne.jp/sutara_lumpur/20120714/1342269933">Blog</a> /
			<a href="http://twitter.com/sutara_lumpur">Twitter</a> /
			<img src="mail_image.png" alt="mail address">
		</address>
	</body>
</html>
